### PR TITLE
chore: document u128 feature

### LINKED ISSUE #4142: Document `u128`
Document https://github.com/noir-lang/noir/pull/3913, more specifically:
- Add descriptions and an example of how it is used since it is different from native uint types (e.g. admonition under https://noir-lang.org/docs/dev/noir/concepts/data_types/integers#signed-integers)
- Shine light on performance hit of using u128 vs other bit sizes, if any

### PR DESCRIPTION
**Pull request by guipublic**
# Description

## Problem\*

Resolves #4142 

## Summary\*

Add documentation for U128

## Documentation\*

Check one:
- [ ] No documentation needed.
- [X] Documentation included in this PR.
- [ ] **[Exceptional Case]** Documentation to be submitted in a separate PR.

# PR Checklist\*

- [X] I have tested the changes locally.
- [X] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

### CODE REVIEW

File: compiler/noirc_evaluator/src/ssa/ir/instruction.rs [hunk 1]
```diff
@@ -476,6 +476,9 @@ impl Instruction {
             }
             Instruction::Truncate { value, bit_size, max_bit_size } => {
                 if let Some((numeric_constant, typ)) = dfg.get_numeric_constant_with_type(*value) {
+                    if bit_size == max_bit_size {
+                        return SimplifiedTo(*value);
+                    }
```
Comment (kevaundray) [inline compiler/noirc_evaluator/src/ssa/ir/instruction.rs:481]: This should be in a separate PR

**Review by kevaundray — APPROVED** [review]

File: docs/docs/noir/concepts/data_types/integers.md [hunk 2]
```diff
@@ -51,6 +51,55 @@ If you are using the default proving backend with Noir, both even (e.g. _u2_, _i
...
+```rust
+fn main(x: u64, y: u64) {
+    let x = U128::from_u64s_be(x,y);
+    assert(z.hi == x as Field);
+    assert(z.lo == y as Field);
+}
+```
+
+Note that the limbs are stored as Field elements in order to avoid unnecessary conversions.
+A part from this, most operations will work as usual:
```
Comment (michaeljklein) [inline docs/docs/noir/concepts/data_types/integers.md:83]: `A part` -> `Apart`

File: docs/docs/noir/concepts/data_types/integers.md [hunk 3]
```diff
@@ -51,6 +51,55 @@ If you are using the default proving backend with Noir, both even (e.g. _u2_, _i
...
+
+```rust
+fn main(x: U128, y: U128) {
+    // multiplication
+    let c = x * y;
+    //addition and subtraction
```
Comment (michaeljklein) [inline docs/docs/noir/concepts/data_types/integers.md:89]: `//addition` -> `// addition`

Comment (guipublic) [reply docs/docs/noir/concepts/data_types/integers.md]: Done

File: docs/docs/noir/concepts/data_types/integers.md (hunk 2 above)
Comment (guipublic) [reply docs/docs/noir/concepts/data_types/integers.md]: Done

**Comment by Savio-Sou** [comment]
cc @noir-lang/developerrelations re doc changes.

File: docs/docs/noir/concepts/data_types/integers.md [hunk 4]
```diff
@@ -51,6 +51,55 @@ If you are using the default proving backend with Noir, both even (e.g. _u2_, _i
 
 :::
 
+
+## 128 bits Unsigned Integers
+
+The built-in structure `U128` allows you to use 128-bit unsigned integers almost like a native integer type. However, there are some differences to keep in mind:
+- You cannot cast between a native integer and `U128`
+- There is a higher performance cost when using `U128`, compared to a native type.
```
Comment (signorecello) [inline docs/docs/noir/concepts/data_types/integers.md:59]: Just a question here @guipublic don't all casts to unsigned integers have a performance cost? I was under that impression (because range checks).

Is this cost simply higher than usual? If not, we should maybe point to the performance cost at the top of the page and remove this one to avoid confusion

Comment (guipublic) [reply docs/docs/noir/concepts/data_types/integers.md:59]: There is no cast for U128, as explained in the doc, so I am not sure to understand your question.
In term of performance, converting an integer to U128 will have no cost (because U128 is the biggest), and converting to a lower bit size will not have a specific drawback because of U128, just what you should expect.

Comment (Savio-Sou) [reply docs/docs/noir/concepts/data_types/integers.md:59]: I guess what @signorecello meant was working with native `uint`s is less performant than working with `Field`s.

This section reads like performance costs would rank as `U128` > native `uint` > `Field`.
Is that the correct way to understand it?

Comment (guipublic) [reply docs/docs/noir/concepts/data_types/integers.md:59]: Yes, the higher the bit size, the higher is the cost. But because U128 uses 2 limbs, cost of arithmetic operations are even higher, especially for multiplication.
